### ISSUE TITLE
Misleading error description, implying that a fat pointer is involved when casting

### ISSUE THREAD

**Issue by Boscop** [rust-lang/rust#43876]
I was writing the following code:
```rust
use glium::glutin::Context;
let win = display.gl_window();
let ctx: &Context = win.context();
let ptr = ctx as *const Context as *mut _ as *mut c_void;
```
It caused the following error:
```
error[E0607]: cannot cast thin pointer `*const glium::<unnamed>::Context` to fat
 pointer `*mut _`
  --> examples\glium.rs:44:12
   |
44 |    let ptr = ctx as *const Context as *mut _ as *mut c_void;
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```
.. even though there is no fat pointer involved.
This error message made me think that [`Context`](https://docs.rs/glium/0.17.0/glium/glutin/struct.Context.html) is a trait (I was looking at [`GlContext`](https://docs.rs/glium/0.17.0/glium/glutin/trait.GlContext.html) before which is a trait) so I ended up boxing the context (`let ptr = Box::into_raw(Box::new(ctx)) as *mut c_void;`).
**The error message could be better, like "type annotation required" or something that doesn't imply that the value is a fat pointer.**

Btw, the point of the above code is to pass the context pointer to libmpv to render video [like this](http://dpaste.com/138XEQ3).

After I realized that Context is not a trait (after trying to figure out how to eliminate the boxing), I changed the code to:
```rust
unsafe extern "C" fn get_proc_address(arg: *mut c_void, name: *const c_char) -> *mut c_void {
	let arg: &Context = &*(arg as *const Context);
	let name = CStr::from_ptr(name).to_str().unwrap();
	arg.get_proc_address(name) as *mut c_void
}

...

let ctx: &Context = win.context();
let ptr = ctx as *const _ as *mut c_void;
```

**Comment by fmease** [comment]
Minimized:

```rs
fn main() {
    struct Context;
    let _ = &Context as *const Context as *mut _;
}
```

Output in Rust 1.19 (2017-07-17):

```
error: cannot cast thin pointer `*const main::Context` to fat pointer `*mut _`
 --> coer.rs:3:13
  |
3 |     let _ = &Context as *const Context as *mut _;
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Current output:

```
error[E0641]: cannot cast to a pointer of an unknown kind
 --> coer.rs:3:43
  |
3 |     let _ = &Context as *const Context as *mut _;
  |                                           ^^^^^^ needs more type information
  |
  = note: the type information given here is insufficient to check whether the pointer cast is valid
```

Closing as fixed.
